upload_events.py

```diff
@@ -61,31 +68,54 @@ def upload_events(file_name: Path, dsn: str, project_id: int, project_slug: str,
         else:
             print(f"Uploading {num_events} events....")
 
-        num_workers = cpu_count() * 2  # empirically 2* num_cpu seems to give the fastest performance
-        q = Queue(num_workers * 2 + 2)  # large enough so that workers do not wait
+        q = Queue(network_threads * 2 + 2)  # large enough so that workers do not wait
 
         workers = []
 
-        for i in range(num_workers):
+        for i in range(network_threads):
             t = Thread(target=worker_loop, args=(dsn, q))
             t.start()
             workers.append(t)
 
         with click.progressbar(events, length=num_events) as events2:
+            last_sleep = time.time()
+            event_count = 0
+
             for event in events2:
-                event.pop('event_id', None)
                 event.pop('project', None)
-                if expire_days:
-                    # Snuba retention is timestamp-based.
-                    event['timestamp'] = now - ((expire_days - 90) * 3600 * 24)
-                else:
-                    event.pop('timestamp', None)
                 event.pop('threads', None)
                 event.pop('debug_meta', None)
-                q.put(event)
+
+                for stacktrace_info in find_stacktraces_in_data(event):
+                    for frame in get_path(stacktrace_info.stacktrace, "frames", filter=True, default=()) or ():
+                        orig_in_app = get_path(frame, "data", "orig_in_app")
+                        if orig_in_app is not None:
+                            frame["in_app"] = None if orig_in_app == -1 else bool(orig_in_app)
+
+                event_id = event.pop('event_id', None)
+                event.setdefault("extra", {})['orig_event_id'] = event_id
+                event['event_id'] = event_id = str(uuid.uuid4().hex).replace("-", "")
+
+                envelope = Envelope(
+                    headers={
+                        "event_id": event_id,
+                        "sent_at": format_timestamp(datetime.utcnow())
+                    }
+                )
+
+                envelope.add_event(event)
+
+                q.put(envelope)
+
+                event_count += 1
+
+                if event_sleep and time.time() - last_sleep > 10 and event_count > 100:
+                    time.sleep(event_count * event_sleep / 1000.0)
+                    last_sleep = time.time()
+                    event_count = 0
```

Comment:
Why do we do this ? (i.e. why do we need to sleep ? )

Reply:
throttling for rate limits. we're asked to send no more than 30 events per second